### PR TITLE
ICU-20700 reimplement acceptLanguage() using the LocaleMatcher; repla…

### PR DESCRIPTION
**Pull request by markusicu**
…ce older accept-language-string parsing by LocalePriorityList

https://unicode-org.atlassian.net/browse/ICU-20700

### CODE REVIEW

**Comment by markusicu** [comment]
> It appears some tests are failing, though.

I had to add one cast to make some compilers happy. Please re-approve.

File: icu4c/source/test/cintltst/cloctst.c [hunk 1]
```diff
@@ -2988,30 +2988,33 @@ static void TestAcceptLanguage(void) {
...
         /*4*/{ 4, NULL, "es", ULOC_ACCEPT_VALID, U_ZERO_ERROR},
-        /*5*/{ 5, NULL, "en", ULOC_ACCEPT_VALID, U_ZERO_ERROR},  /* XF */
+        /*5*/{ 5, NULL, "zh", ULOC_ACCEPT_FALLBACK, U_ZERO_ERROR},  /* XF */
         /*6*/{ 6, NULL, "ja", ULOC_ACCEPT_FALLBACK, U_ZERO_ERROR},  /* XF */
         /*7*/{ 7, NULL, "zh", ULOC_ACCEPT_FALLBACK, U_ZERO_ERROR},  /* XF */
-        /*8*/{ 8, NULL, "", ULOC_ACCEPT_FAILED, U_ZERO_ERROR },  /*  */
-        /*9*/{ 9, NULL, "", ULOC_ACCEPT_FAILED, U_ZERO_ERROR },  /*  */
-       /*10*/{10, NULL, "", ULOC_ACCEPT_FAILED, U_BUFFER_OVERFLOW_ERROR },  /*  */
-       /*11*/{11, NULL, "", ULOC_ACCEPT_FAILED, U_BUFFER_OVERFLOW_ERROR },  /*  */
+        /*8*/{ 8, NULL, "", ULOC_ACCEPT_FAILED, U_ILLEGAL_ARGUMENT_ERROR },  /*  */
```
[the remark this answers is not in the document]
Comment (markusicu) [reply icu4c/source/test/cintltst/cloctst.c]: Seems to me like an edge case that neither needs to be stable nor mentioned.

File: icu4c/source/test/cintltst/cloctst.c [hunk 2]
```diff
@@ -2988,30 +2988,33 @@ static void TestAcceptLanguage(void) {
...
-        /*3*/ "mga-ie;q=0.9, tlh",
-        /*4*/ "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "
-              "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "
-              "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "
-              "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "
-              "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "
-              "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "
-              "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xx-yy;q=.1, "
+        /*3*/ "mga-ie;q=0.9, sux",
+        /*4*/ "xxx-yyy;q=0.01, xxx-yyy;q=0.01, xxx-yyy;q=0.01, xxx-yyy;q=0.01, xxx-yyy;q=0.01, "
```
[the remark this answers is not in the document]
Comment (markusicu) [reply icu4c/source/test/cintltst/cloctst.c]: The Java LocalePriorityList (which is public API) has been there for a few years. I made it more efficient but didn't really change its behavior.

When I wrote a C++ version of that last summer (not for public API for now), I looked at the RFC for the syntax, and implemented q number parsing accordingly.

Both the Java LocalePriorityList and the previous uloc_acceptLanguage() code used generic number parsing functions. Yes, they are now inconsistent. Java is still more lenient. I don't know whether this matters to anyone -- depends on whether someone strongly needs us to be able to parse leniently. (Like we do parse locale IDs quite leniently.)

If you think this should be revisited, then please submit a ticket, explain your use case, and make a recommendation.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/util/ULocaleTest.java [hunk 3]
```diff
@@ -1590,15 +1589,15 @@ private void initHashtable() {
...
+        /*7*/ { "zh_Hant", "true" },
     };
 
     private static final String ACCEPT_LANGUAGE_HTTP[] = {
         /*0*/ "mt-mt, ja;q=0.76, en-us;q=0.95, en;q=0.92, en-gb;q=0.89, fr;q=0.87, iu-ca;q=0.84, iu;q=0.82, ja-jp;q=0.79, mt;q=0.97, de-de;q=0.74, de;q=0.71, es;q=0.68, it-it;q=0.66, it;q=0.63, vi-vn;q=0.61, vi;q=0.58, nl-nl;q=0.55, nl;q=0.53, th-th-traditional;q=.01",
         /*1*/ "ja;q=0.5, en;q=0.8, tlh",
         /*2*/ "en-zzz, de-lx;q=0.8",
-        /*3*/ "mga-ie;q=0.9, tlh",
+        /*3*/ "mga-ie;q=0.9, sux",
         /*4*/ "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "+
```
[the remark this answers is not in the document]
Comment (markusicu) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/util/ULocaleTest.java]: This _is_ Java land. CLDR languageMatch data has a one-way fallback from Klingon to English, so I changed the test case to use Sumerian which still does not match anything else.

[the remark this answers is not in the document]
Comment (markusicu) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/util/ULocaleTest.java]: Or are you talking about the q number parsing? For that, see above.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/util/ULocaleTest.java [hunk 4]
```diff
@@ -1610,145 +1609,67 @@ private void initHashtable() {
                 "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "+
                 "xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, xxx-yyy;q=.01, "+
                 "es",
-                /*5*/ "de;q=.9, fr;q=.9, xxx-yyy, sr;q=.8",
-                /*6*/ "zh-tw",
-                /*7*/ "zh-hant-cn",
+        /*5*/ "de;q=.9, fr;q=.9, xxx-yyy, sr;q=.8",
```
[the remark this answers is not in the document]
Comment (markusicu) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/util/ULocaleTest.java]: We have over 800 test cases for the C++ LocaleMatcher, and acceptLanguage() is now a thin wrapper over that. Adding lots of test cases for the wrapper seems pointless.
